clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/DefaultIngestionBackend.java

```diff
@@ -247,52 +252,71 @@ private void adjustStoragePartitionWhenBlobTransferComplete(StorageEngine storag
+   * This is because retained data may not be cleaned up unless a new host is added, making it difficult to validate this feature.
+   * This 'blobTransferDisabledOffsetLagThreshold' config ensures blob transfer always runs in such cases.
    *
    * @param store the store name
    * @param versionNumber the version number
    * @param partition the partition number
    * @param blobTransferDisabledOffsetLagThreshold the maximum allowed offset lag threshold.
+   *        This value is controlled by config BLOB_TRANSFER_DISABLED_OFFSET_LAG_THRESHOLD, and default is 100000L.
    *        If the offset lag is within this threshold, bootstrapping from Kafka is allowed, even if blob transfer is enabled.
    *        If the lag exceeds this threshold, bootstrapping should happen from blobs transfer firstly.
-   *
-   * @return true if the offset lag exceeds the threshold or if the lag is 0, indicating bootstrapping should happen from blobs transfer.
-   *         false otherwise
+   * @param hybridStore whether the store is a hybrid store or not.
+   *                    If it is a hybrid store, then check via the offset.
+   *                    If it is a batch store, check if the batch push is done or not.
+   * @return true if the store is lagged and needs to bootstrap from blob transfer, else false then bootstrap from Kafka.
    */
   public boolean isOffsetLagged(
       String store,
       int versionNumber,
       int partition,
-      long blobTransferDisabledOffsetLagThreshold) {
+      long blobTransferDisabledOffsetLagThreshold,
+      boolean hybridStore) {
     String topicName = Version.composeKafkaTopic(store, versionNumber);
     OffsetRecord offsetRecord = storageMetadataService.getLastOffset(topicName, partition);
 
-    if (offsetRecord == null || (offsetRecord.getOffsetLag() == 0 && offsetRecord.getLocalVersionTopicOffset() == -1)) {
-      LOGGER.info(
-          "Offset record is null or offset lag is 0 and topic offset is -1 for store {} partition {}.",
-          store,
-          partition);
+    if (offsetRecord == null) {
       return true;
     }
 
-    if (offsetRecord.getOffsetLag() < blobTransferDisabledOffsetLagThreshold) {
-      LOGGER.info(
-          "Offset lag {} for store {} partition {} is within the allowed lag threshold {}. Bootstrapping from Kafka.",
-          offsetRecord.getOffsetLag(),
-          store,
-          partition,
-          blobTransferDisabledOffsetLagThreshold);
-      return false;
+    if (blobTransferDisabledOffsetLagThreshold < 0) {
+      return true;
+    }
+
+    if (!hybridStore) {
+      if (offsetRecord.isEndOfPushReceived()) {
+        LOGGER.info(
+            "End of push received for batch store replica {}, might due to restore. Bootstrapping from Kafka.",
+            Utils.getReplicaId(topicName, partition));
+        return false;
+      }
+    } else {
+      if (offsetRecord.getOffsetLag() == 0 && offsetRecord.getLocalVersionTopicOffset() == -1) {
```

Comment:
`offsetRecord.getOffsetLag()` will only return the previous persisted offset lag, which may not reflect the current lag.
For example, before restart, the previous offset lag is minimal, and the restart takes 20 mins, and the new lag can be very large for some high write-throughput use cases, and this check won't be able to catch that.
I think for offset lag check, we need a real-time check rather than the previous-persisted offset lag.